Convert Checkup output to SARIF (#2)
checkupjs/rfcs: text/0002-sarif-result-schema.md

- Feature Name: Migrating Result Schema to SARIF 
- Start Date: 2020-09-01
- RFC PR:
- Checkup Issue: 

# Summary
[summary]: #summary

In [a previous RFC](./0001-task-result-schema.md), the need to standardize the output of Checkup's data was discussed. In this RFC, I propose leveraging the [SARIF](https://github.com/microsoft/sarif-tutorials) schema to shape the Checkup output. 

# Motivation
[motivation]: #motivation

> Since the [Task Result Schema RFC](./0001-task-result-schema.md), changes have been implemented to the output that align with the schema described in that RFC, cleaning up the output significantly and adding shared helpers that help task authors format their data properly to fit the schema. Though defining a refined schema was a useful step in structuring the output, it didn't leverage any standardized schema for static analysis, such as the SERIF format from Microsoft.

Reviewer: ```suggestion
Since the [Task Result Schema RFC](./0001-task-result-schema.md), changes have been implemented to the output that align with the schema described in that RFC, cleaning up the output significantly and adding shared helpers that help task authors format their data properly to fit the schema. Though defining a refined schema was a useful step in structuring the output, it didn't leverage any standardized schema for static analysis, such as the SARIF format from Microsoft.
```

> SARIF is a well-established schema that "defines a standard format for the output of static analysis tools." There are a multitude of tools and plugins available in the open-source community that leverage SARIF - aligning our data structure with this shared schema will allow the data produced by checkup to be consumed by data pipelines that are already consuming SARIF data. For example, there is a [VSCode plugin](https://github.com/microsoft/sarif-vscode-extension) that prettyprints SARIF data, a [react based visualization tool](https://github.com/Microsoft/sarif-web-component), and a package that [converts eslint data to SARIF](https://www.npmjs.com/package/%40microsoft%2Feslint-formatter-sarif

Reviewer 2: ```suggestion
SARIF is a well-established schema that "defines a standard format for the output of static analysis tools." There are a multitude of tools and plugins available in the open-source community that leverage SARIF - aligning our data structure with this shared schema will allow the data produced by checkup to be consumed by data pipelines that are already consuming SARIF data. For example, there is a [VSCode plugin](https://github.com/microsoft/sarif-vscode-extension) that prettyprints SARIF data, a [react based visualization tool](https://github.com/Microsoft/sarif-web-component), and a package that [converts eslint data to SARIF](https://www.npmjs.com/package/%40microsoft%2Feslint-formatter-sarif)
```

      * 
      * CHECKUP: this will generally be just one Run
      */
      runs: Run[];

      /**
      * Key/value pairs that provide additional information about the physical location.
      *
      * CHECKUP: we can store the project info, analyzedFiles, actions, and analyzedFilesCount here
      *
      */
      properties?: PropertyBag;

  }

  export interface Run {
      /**
      * Describes the invocation of the analysis tool. 
      *
      * CHECKUP: we will put cli info here
      */
      invocations?: Invocation[];

      
      /**
      * The set of results contained in an SARIF log. The results array can be omitted when a run is solely exporting
      * rules metadata. It must be present (but may be empty) if a log file represents an actual scan.
      */
      results?: Result[];

      /**
      * Information about the tool or tool pipeline that generated the results in this run. A run can only contain
      * results produced by a single tool or tool pipeline. A run can aggregate results from multiple log files, as long
      * as context around the tool run (tool command-line arguments and the like) is identical for all aggregated files.
      *

>       * CHECKUP: We may not _need_ to populate this, but I don't see why not :) There is a use case for products using this schema for other 

Reviewer 2: We should definitely populate this. This is to be consistent with the SARIF spec, but also to clearly disambiguate the source of the data (from which tool it came).

Author: Cool, sg - I am populating it in my implementation, but I will modify this comment

      */
      releaseDateUtc?: string;

      /**
      * An array of reportingDescriptor objects relevant to the analysis performed by the tool component.
      */
      rules?: ReportingDescriptor[];

      /**
      * The tool component version in the format specified by Semantic Versioning 2.0.
      */
      semanticVersion?: string;

      /**
      * A brief description of the tool component.
      */
      shortDescription?: MultiformatMessageString;

      /**
      * The tool component version, in whatever format the component natively provides.
      */
      version?: string;

      /**
      * Key/value pairs that provide additional information about the tool component.
      */
      properties?: PropertyBag;
  }

  /**
  * The runtime environment of the analysis tool run.
  */
  export interface Invocation {
      /**
      * An array of strings, containing in order the command line arguments passed to the tool from the operating
      * system.
      *
      * CHECKUP: args go heree 
      */
      arguments?: string[];

      /**
      * The command line used to invoke the tool.
      */

>       commandLine?: string;

Reviewer 2: Is this the full command line string that's used to invoke the tool?

Author: Yes

      /**
      * The Coordinated Universal Time (UTC) date and time at which the run ended. See "Date/time properties" in the
      * SARIF spec for the required format.
      * 
      * CHECKUP: this property along with startTimeUtc will contain timing info
      */
      endTimeUtc?: string;

      /**
      * The environment variables associated with the analysis tool process, expressed as key/value pairs.
      * 

>       * CHECKUP: this property will contain the flags, config, configHash 

Reviewer 2: Do they explicitly mean env vars only? Are we overloading this with putting flags/config, etc in?

Reviewer 2: Additionally, oclif provides a [boatload of env var details](https://oclif.io/docs/config) when the CLIs are run. I believe they're in the `config` object (`this.config` in a command).

Author: They _may_ explicitly mean env vars - the spec wasnt super clean imo. 

```
An invocation object MAY contain a property named environmentVariables whose value is an object. The property names in this object SHALL contain the names of all the environment variables in the tool's execution environment. The value of each property SHALL be a string containing the value of the specified environment variable. If the value of the environment variable is an empty string, the corresponding property value SHALL be an empty string.

NOTE 1: Environment variable names and values are likely to reveal highly sensitive information. For example, on a Windows machine, environment variables reveal the directories on the execution path, user account name, machine name, logon domain controller, etc.

NOTE 2: The result of setting an environment variable to an empty string is operating system-dependent. On Windows, it removes the variable from the environment. In Unix, an environment variable can have an empty value.

```
 

What do you think @scalvert ?

Reviewer 2: I think the 'may' part was referring to the fact that `environmentVariables` may be present or not, not whether it may contain env vars or not. I think we'd be overloading this if we used it.

      */
      environmentVariables?: { [key: string]: string };

      /**
      * Specifies whether the tool's execution completed successfully.
      */
      executionSuccessful: boolean;

      /**
      * The process exit code.
      */
      exitCode?: number;

      /**
      * The machine that hosted the analysis tool run.
      */

>       machine?: string;

Reviewer 2: I think https://oclif.io/docs/config contains this info.

Author: ```
An invocation object MAY contain a property named machine whose value is a string containing the name of the machine on which the tool was run.
``` 

It seems like oclif has the name of the platform and the shell, but I don't see the name of the machine?

Reviewer 2: Got it.

      /**
      * The Coordinated Universal Time (UTC) date and time at which the run started. See "Date/time properties" in the
      * SARIF spec for the required format.
      */
      startTimeUtc?: string;

      /**
      * A list of runtime conditions detected by the tool during the analysis.
      * 

>       * CHECKUP: this is where we will store nonterminal errors collected from Tasks in checkup

Reviewer 2: I like!

Author: <3

      */
      toolExecutionNotifications?: Notification[];

  }

  export namespace Result {
      type kind =
          "notApplicable" |
          "pass" |
          "fail" |
          "review" |
          "open" |
          "informational";

      type level =
          "none" |
          "note" |
          "warning" |
          "error";
  }

  /**
  * A result produced by an analysis tool.
  */
  export interface Result {
      /**
      * An array of 'fix' objects, each of which represents a proposed fix to the problem indicated by the result.
      * 
      */
      fixes?: Fix[];

      /**
      * A stable, unique identifer for the result in the form of a GUID.
      */
      guid?: string;
      
      /**
      * A value that categorizes results by evaluation state.
      */

>       kind?: Result.kind;

Reviewer 2: This field makes me think - I've been looking to add the ability for you to specify whether you want a task to be able to raise and error (errorCode != 0) in order for us to be able to use checkup for validations, for example

- correct lint configuration structure
- correct project configuration structure (think correct Volta setup)

This could be used to populate that?

Author: ```
    type kind =
        "notApplicable" |
        "pass" |
        "fail" |
        "review" |
        "open" |
        "informational";
```  

This could apply? I think this can be added as a feature in a follow up for sure! Good idea. If you think these types apply for what you are trying to do, lmk and i can create an issue to build that out

Reviewer 2: We're no longer going to add the `errorCode` feature in this way, so we can ignore this!

      /**
      * A value specifying the severity level of the result.
      */
      level?: Result.level;

      /**
      * A message that describes the result. The first sentence of the message only will be displayed when visible space
      * is limited.
      *

>       * CHECKUP: if we wanted to keep concepts of Result type `SummaryResult | MultiValueResult | LookupValueResult;`, we can store those here, as well as the dataSummary and data for each Result

Reviewer 2: Good suggestion.

      "config": {
        "plugins": [
          "checkup-plugin-ember",
          "checkup-plugin-ember-octane",
          "checkup-plugin-javascript"
        ],
        "tasks": {
          "ember/ember-test-types": [
            "on",
            {
              "actions": {
                "ratioApplicationTests": {
                  "threshold": 3
                }
              }
            }
          ]
        }
      },
      "version": "0.7.4",
      "schema": 1,
      "args": {
        "paths": []
      },
      "flags": {
        "format": "json",
        "outputFile": ""
      },
      "timings": {
        "ember/ember-types": 23.371086841,
        "ember/ember-in-repo-addons-engines": 23.329550989,
        "ember/ember-dependencies": 23.327852732,
        "javascript/eslint-summary": 20.294677398,
        "ember/ember-test-types": 23.346301169,
        "ember-octane/ember-octane-migration-status": 8.863200936,
        "ember/ember-template-lint-disables": 27.640971204,
        "javascript/eslint-disables": 17.036571088,
        "javascript/outdated-dependencies": 43.921809849
      }
    },
    "analyzedFiles": [],
    "analyzedFilesCount": 1539,

>     "actions": [

Reviewer 2: Is there a more "SARIF"-style location for actions?

Author: So I don't think so. I have tried, specifically tried moving the actions to be stored with the Results because that makes more sense with SARIF's mental model, but I couldn't find anywhere to put them. Do you have any ideas?

Reviewer 2: No, it was more of a question. I got nothing :(

          "message": {
            "text": "components"
          },
          "locations": [],
          "occurrenceCount": 219,
          "properties": {
            "taskName": "ember-types",
            "taskDisplayName": "Ember Types",
            "category": "metrics",
            "group": "ember"
          }
        },
        {
          "message": {
            "text": "controllers"
          },
          "locations": [],
          "occurrenceCount": 41,
          "properties": {
            "taskName": "ember-types",
            "taskDisplayName": "Ember Types",
            "category": "metrics",
            "group": "ember"
          }
        },
        {
          "message": {
            "text": "helpers"
          },
          "locations": [],
          "occurrenceCount": 35,
          "properties": {
            "taskName": "ember-types",
            "taskDisplayName": "Ember Types",
            "category": "metrics",
            "group": "ember"
          }
        },
        {
          "message": {
            "text": "initializers"
          },
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {

>                   "uri": "/app/initializers/app.js"

Reviewer 2: Are these paths meant to be absolute? This looks vaguely wrong...

It should be `app/initializers/app.js` no?

It's possible this was my mistake in how the output currently is.

Author: Yeah I think there shouldn't  be the initial forward slash there.. I will make sure to fix in the impl, and will correct all paths here

                }
              }
            },
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "/app/initializers/configure-inflector.js"
                }
              }
            },
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "/app/initializers/services.js"
                }
              }
            },
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "/app/initializers/tracer.js"
                }
              }
            }
          ],
          "occurrenceCount": 4,

>           "properties": {

Reviewer 2: The challenge with stuffing things into a property bag like this is you need intimate knowledge of the contents of the property bag in order to access it. It's not easily iterable, and it's hard to write dynamic loops that try to generically access this data. WDYT?

Author: We discussed this a bit offline; I agree with the risks/problems with this, but I think the pros of using this schema as a whole outweigh the cons, and I don't see another option for this type of  data with SARIF. WDYT?

Reviewer 2: I'd tend to agree. That's the price we're paying with using a standardized format - we have to accommodate the fact that it's general purpose.

- Feature Name: Migrating Result Schema to SARIF 
- Start Date: 2020-09-01
- RFC PR:
- Checkup Issue: 

# Summary
[summary]: #summary

In [a previous RFC](./0001-task-result-schema.md), the need to standardize the output of Checkup's data was discussed. In this RFC, I propose leveraging the [SARIF](https://github.com/microsoft/sarif-tutorials) schema to shape the Checkup output. 

# Motivation
[motivation]: #motivation

> Since the [Task Result Schema RFC](./0001-task-result-schema.md), changes have been implemented to the output that align with the schema described in that RFC, cleaning up the output significantly and adding shared helpers that help task authors format their data properly to fit the schema. SARIF is a well-established schema that "defines a standard format for the output of static analysis tools." There are a ton of tools and plugins available in the open-source community that leverage SARIF - aligning our data structure with this shared schema will allow the data produced by checkup to be consumed by data pipelines that are already consuming SARIF data. For example, there is a [VSCode plugin](https://github.com/microsoft/sarif-vscode-extension) that prettyprints SARIF data, a [react based visualization tool](https://github.com/Microsoft/sarif-web-component), and a package that [converts eslint data to SARIF](https://www.npmjs.com/package/%40microsoft%2Feslint-formatter-sarif

Reviewer 2: ```suggestion
Since the [Task Result Schema RFC](./0001-task-result-schema.md), changes have been implemented to the output that align with the schema described in that RFC, cleaning up the output significantly and adding shared helpers that help task authors format their data properly to fit the schema. Though defining a refined schema was a useful step in structuring the output, it didn't leverage any standardized schema for static analysis, such as the SERIF format from Microsoft.

SARIF is a well-established schema that "defines a standard format for the output of static analysis tools." There are a multitude of tools and plugins available in the open-source community that leverage SARIF - aligning our data structure with this shared schema will allow the data produced by checkup to be consumed by data pipelines that are already consuming SARIF data. For example, there is a [VSCode plugin](https://github.com/microsoft/sarif-vscode-extension) that prettyprints SARIF data, a [react based visualization tool](https://github.com/Microsoft/sarif-web-component), and a package that [converts eslint data to SARIF](https://www.npmjs.com/package/%40microsoft%2Feslint-formatter-sarif
```

Author: accepted suggestion

). Additionally, developers familiar with SARIF will have an easier time onboarding to Checkup. 
  

> # Details

Reviewer 2: Can we also add an example of pre and post migrated JSON output, so we can visualize what the changes will look like?

Author: adding now!

[details]: #details

This migration will be pretty straightforward for the following reasons:
- SARIF schema isn't too far off from the schema we have already implemented to return result data. 
- The creation utilties used by Tasks to create CheckupResult data are abstracted cleanly from the Tasks themselves, so we can make this change without having to really touch the Tasks much at all. 

Here is are the current types for structuring result data: 

> ```  typescript

Reviewer 2: You an wrap these in

```
<details>
  <summary>foo</summary>
  //...
</details>
```

which will collapse them by default.

export interface CheckupResult {
  info: {
    project: {
      name: string;
      version: string;
      repository: {
        totalCommits: number;
        totalFiles: number;
        age: string;
        activeDays: string;
      };
    };
    cli: {
      schema: number;
      configHash: string;
      config: CheckupConfig;
      version: string;
      args: {
        paths: string[];
      };
      flags: {
        config?: string;
        task?: string[];
        format: string;
        outputFile?: string;
        excludePaths?: string[];
      };
      timings: Record<string, number>;
    };
    analyzedFiles: string[];
    analyzedFilesCount: number;
  };
  results: TaskResult[];
  errors: TaskError[];
  actions: Action[];
}
```

With SARIF, the results would look like this (there are a ton of optional properties here, so I am keeping only the ones we will need): 
*Note* I left in the comments from SARIF because I think they are useful to understand the shape of the data here, so when I wanted to add 
checkup related commentary I added a newline after the existing SARIF comments and prefaced with `CHECKUP:` before adding checkup commentary to differentiate.

> ```  typescript

Reviewer 2: You an wrap these in

```
<details>
  <summary>foo</summary>
  //...
</details>
```

which will collapse them by default.

     * A plain text message string.
     */
    text?: string;

    /**
     * Key/value pairs that provide additional information about the message.
     */
    properties?: PropertyBag;
}

/**
 * A proposed fix for the problem represented by a result object. A fix specifies a set of artifacts to modify. For
 * each artifact, it specifies a set of bytes to remove, and provides a set of new bytes to replace them.
 */
export interface Fix {
    /**
     * One or more artifact changes that comprise a fix for a result.
     */
    artifactChanges: ArtifactChange[];

    /**
     * A message that describes the proposed fix, enabling viewers to present the proposed change to an end user.
     */
    description?: Message;

    /**
     * Key/value pairs that provide additional information about the fix.
     */
    properties?: PropertyBag;
}

```
Note: there are type definitions for SARIF [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/289ba098a611b80ee03ed72a2feee41a593fe765/types/sarif/index.d.ts). These types are being used by a number of the high profile tools I linked above, so I view them as very reliable, even though they are third party.

## Open questions: 

> 1) Where is it best to store TaskErrors? Options are: properties on `Invocation`, properties on `Run`. I don't see a more specific place in SARIF schema to put nonterminal errors

Reviewer 2: I think errors should be stored in the `Run` propertyBag. This will keep it on-par with the existing schema, where the errors property is a sibling of the results.

Reviewer 3: Are `TaskErrors` "meta" errors (failures of the task itself rather than the code)? If so, then `invocation.toolExecutionNotifications` seems like the right place for them to live rather than any custom property bag. See the [SARIF tutorials](https://github.com/microsoft/sarif-tutorials/blob/master/docs/3-Beyond-basics.md#tool-execution-notification-and-tool-configuration-notifications)

Reviewer 2: They are. This is a great suggestion, @drewhannay. 

@carakessler, let's make that change.

Author: Good call - thank you @drewhannay!

> 2) Are there concerns with storing `actions` as `fixes` per each `Result`, instead of aggregating all `actions` a level up  (on the `Run` itself)?

Reviewer 2: This seems like a good change, and will allow us to align with SERIF's format.

Author: @scalvert - in trying to implement this, I realized that `Fix` is not really an appropriate construct to cleanly shove `Actions` into, as it requires [ArtifactChange](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/289ba098a611b80ee03ed72a2feee41a593fe765/types/sarif/index.d.ts#L201) and [ArtifactLocation](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/289ba098a611b80ee03ed72a2feee41a593fe765/types/sarif/index.d.ts#L247) which _could_ apply to some actions, but not cleanly to all at this time. I am going to leave `Actions` in the `PropertyBag` for initial migration, and then re-evaluate if we want change how we think about Actions to align closer with the SARIF construct of Fixes
